concrete/dense_directed_matrix.go

```diff
@@ -112,8 +115,16 @@ func (g *DirectedDenseGraph) HasEdgeFromTo(u, v graph.Node) bool {
 	return uid != vid && !isSame(g.mat.At(uid, vid), g.absent)
 }
 
-func (g *DirectedDenseGraph) Weight(e graph.Edge) float64 {
-	return g.mat.At(e.From().ID(), e.To().ID())
+func (g *DirectedDenseGraph) Weight(x, y graph.Node) (w float64, ok bool) {
+	xid := x.ID()
+	yid := y.ID()
+	if xid == yid {
+		return 0, true
```

Comment:
>Sorry for unnecessary questions, I got to the Weighter interface in the meantime.

Not unnecessary at all, they go to the heart of a matter that troubles me.

>What other reasonable values are there for absent apart from Inf(1)? Accepting that any number, then 1/absent feels a bit arbitrary.

Sane values for absent IMO are 0 (correlation graph), Inf (cost graph) and NaN (dubious but arguable). This gives the plausible behaviour for the cost graph case (a correlation graph would have self edges being 1 and I don't know what someone who is claiming an absent edge == NaN would want a self edge to be, it could be either of 0, 1 or Inf). Maybe there should be two parameters - self and absent. I don't like this very much, but maybe it's the best of a bad lot.

>At any case, for completeness, shouldn't there be an (unlikely, but possible) subcase that `xid == yid && !g.has(xid)` ?

Not sure. The cost of an edge to itself is alway 0 and the correlation is always 1. This is independent of whether the graph knows about it. We can add it though.

I'm up for suggestions on this. We provide these graphs, and they should be reasonably generalisable, but covering all the cases opens us to configurations that are internally inconsistent in client code. And we bear the support cost.